### PR TITLE
[core][state] Optimize task event buffer w/o merging TaskEvents

### LINKED ISSUE #31546: [core][state] Microbench regression with task backend investigation
### What happened + What you expected to happen

image

Investigate further if we could resolve the regression or identify the root cause (which function call/callsite causes this) 

### Versions / Dependencies

master

### Reproduction script

NA

### Issue Severity

None

### PR DESCRIPTION
**Pull request by rickyyx**
Signed-off-by: rickyyx <rickyx@anyscale.com>

## Why are these changes needed?
Since work in the separate thread in CoreWorker will potentially impact the performance of ray if as many as workers are started by default, we need to reduce as much work as possible on the worker side. 

This PR skips merging of task events on the worker side and conversion from intermediate data structures to the sendable `rpc::TaskEventData`

## Related issue number
Closes https://github.com/ray-project/ray/issues/31546

⚠️  **NOTE: this doesn't actually bring back the regression since there will still be contention with the microbenchmark setup.**  

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/task_event_buffer.cc [hunk 1]
```diff
@@ -100,98 +102,85 @@ void TaskEventBufferImpl::AddTaskEvent(rpc::TaskEvents task_events) {
...
     // Too many task events, start overriding older ones.
-    if (buffer_[next_idx_to_overwrite_].has_profile_events()) {
+    if (data_->events_by_task().at(next_idx_to_overwrite_).has_profile_events()) {
       num_profile_task_events_dropped_++;
     } else {
       num_status_task_events_dropped_++;
     }
 
-    buffer_[next_idx_to_overwrite_] = std::move(task_events);
+    // The two should be from different arena : so calling Swap will not improve
```
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.cc:113]: what do you mean by "the two"?

File: src/ray/core_worker/task_event_buffer.cc [hunk 2]
```diff
@@ -100,98 +102,85 @@ void TaskEventBufferImpl::AddTaskEvent(rpc::TaskEvents task_events) {
...
 }
 
 void TaskEventBufferImpl::FlushEvents(bool forced) {
   if (!enabled_) {
     return;
   }
-  std::vector<rpc::TaskEvents> task_events;
-  size_t num_status_task_events_dropped = 0;
-  size_t num_profile_task_events_dropped = 0;
+  std::unique_ptr<rpc::TaskEventData> data_to_send =
```
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.cc:127]:
```suggestion
  auto data_to_send =
```

**Review by rkooo567 — APPROVED** [review]
So we just store data to protobuf instead of in memory vector? How much improvement does it make?

**Comment by rickyyx** [comment]
> So we just store data to protobuf instead of in memory vector? How much improvement does it make?

Not much on the CD microbenchmark since there's still contention, but it improves by 5% (although not stable) from local microbenchmark when I limit the number of cores used by the ray cluster.

File: src/ray/core_worker/task_event_buffer.cc (hunk 1 above)
Comment (rickyyx) [reply src/ray/core_worker/task_event_buffer.cc:113]: The new task_events and the one to be overwritten.

**Comment by rickyyx** [comment]
Hmmm, seems it worsens the CD microbenchmark performance on some of the metrics. Given this doesn't really make things much better, I will not merge this.

**Pull request closed without merging**